playground/src/components/details/editor/editor_widget.tsx

```diff
@@ -0,0 +1,97 @@
+import { useEffect, useId, useMemo, useRef } from "react";
+import { createRoot } from "react-dom/client";
+import { editor } from "monaco-editor";
+import { createHtmlPortalNode, InPortal, OutPortal } from "react-reverse-portal";
+
+type EditorWidgetProps = {
+  editor: editor.IStandaloneCodeEditor;
+  children: React.ReactNode;
+  afterLineNumber: number;
+  afterColumn?: number;
+  modelPath?: string;
+};
+
+/**
+ * A editor widget that can be placed inline in the Monaco text editor.
+ * 
+ * We render the widget in a portal so it can be a React component using react state etc.
+ * The out-portal is rendered in the editor as a overlay widget and a view zone makes space for it.
+ */
+export function EditorWidget({ editor, children, afterLineNumber, afterColumn, modelPath }: EditorWidgetProps) {
+  const id = useId();
+  const portalNode = useMemo(() => createHtmlPortalNode(), []);
+  const resizeObserverRef = useRef<ResizeObserver>();
+  const overlayWidgetRef = useRef<editor.IOverlayWidget>();
+  const zoneIdRef = useRef<string>();
+  
+  const updateWidget = () => {
+    overlayWidgetRef.current && editor.removeOverlayWidget(overlayWidgetRef.current);
+    zoneIdRef.current && editor.changeViewZones((accessor) => accessor.removeZone(zoneIdRef.current!));
```

Comment:
```suggestion
    if (overlayWidgetRef.current) {
        editor.removeOverlayWidget(overlayWidgetRef.current);
    }
    if (zoneIdRef.current) {
        editor.changeViewZones((accessor) => accessor.removeZone(zoneIdRef.current));
    }
```
(I don't know if omitting the `!` in `zoneIdRef.current!` really works then in TS)

Reply:
Changed. JS supports ! as far as I know, for example you can so `var?.someFunc()`, so it also makes sense to have `var!.someFunc()` which should throw an error AFAIK

Reply:
That won't throw an error; it will only tell TS to shut up about type errors ;) https://stackoverflow.com/a/42274019/4306257 

Therefore, it has no effect here.

Reply:
However, we get a type error here otherwise -> we need the `!`